### PR TITLE
Improve the dot decoration placement as we edit

### PR DESCRIPTION
**Pull request by seanpoulter**
The dot decorations would often appear in the wrong spot as we edit the document because they updated after every change using locations from the last results.   
<br/>

This commit stops updating the test result decorations when there are unsaved changes to the document. Using the `onDidChangeTextDocument` event handler, we can detect when the document is saved or updated by an external editor and update accordingly. It also is fired when other Resources are changed (e.g.: using the `git` scheme from the Git extension).

To keep the decorations in place as we edit or format the document, the the decoration types are now created with a `DecorationRangeBehavior` that prevents the range from growing and gives them a "sticky" feel.

This commit also changes the way the inline error messages are cached to keep the error messages visible in all instances of a document, not just in the active editor.
<br/>

This addresses most of #116 but there are at least two problems:
* the Debug CodeLens is in the wrong spot, and the API doesn't seem to support range "stickiness"
* if you open the editor to a document with unsaved changes, the dots appear in the wrong spot

Both of those would be solved with a language service that keeps track of the test positions and results for unsaved changes.

### CODE REVIEW

File: src/JestExt.ts [hunk 1]
```diff
@@ -545,4 +552,47 @@ export class JestExt {
...
     this.testResultProvider.removeCachedResults(filePath)
   }
+
+  removeCachedDecorationTypes(document: vscode.TextDocument) {
+    if (!document || !document.fileName) {
+      return
+    }
+
+    delete this.failingAssertionDecorators[document.fileName]
+  }
```
Comment (orta) [inline src/JestExt.ts]: nice last week I almost made a PR for all this - https://github.com/jest-community/vscode-jest/tree/turn_off_stuff

Comment (orta) [reply src/JestExt.ts]: well, the removal anyway

File: src/decorations.ts [hunk 2]
```diff
@@ -16,6 +16,7 @@ export function failingItName() {
         contentText: '●',
       },
     },
+    rangeBehavior: DecorationRangeBehavior.ClosedClosed,
```
Comment (orta) [inline src/decorations.ts]: so closed they wrote it twice

File: tests/extension.test.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    it('should register an event handler to handle when a document is saved', () => {
+      const handler = jestInstance.onDidChangeTextDocument
+      activate(context)
+
+      expect(vscode.workspace.onDidChangeTextDocument).toBeCalledWith(handler, thisArg)
+      expect(context.subscriptions.push).toBeCalled()
+      expect(context.subscriptions.push.mock.calls[0]).toContain('onDidChangeTextDocument')
+    })
+  })
+})
```
Comment (orta) [inline tests/extension.test.ts]: damn! This is a lot of tests, congrats

**Comment by orta** [comment]
TBH, converting jest-editor-support to be a language server would be an epic move

**Comment by orta** [comment]
Shipped as 2.5.6 👍 

Nice work, we should get you on the README somewhere as a maintainer

File: src/decorations.ts (hunk 2 above)
Comment (seanpoulter) [reply src/decorations.ts]: It's a terrible name, and unfortunately they haven't added the legible [alias for it](https://github.com/Microsoft/vscode/blob/114adf703be6203fba2da696aeac439449241ac9/src/vs/editor/common/editorCommon.ts#L899) to the enum:
```
export enum TrackedRangeStickiness {
	AlwaysGrowsWhenTypingAtEdges = 0,
	NeverGrowsWhenTypingAtEdges = 1,
	GrowsOnlyWhenTypingBefore = 2,
	GrowsOnlyWhenTypingAfter = 3,
}
```

The default is `OpenOpen` or `AlwaysGrowsWhenTypingAtEdges`.

File: src/JestExt.ts (hunk 1 above)
Comment (seanpoulter) [reply src/JestExt.ts]: Nice. Assign me to the review if you get **that** annoyed with them again. :wink:

**Comment by seanpoulter** [comment]
I've created Issue #205 to discuss the language server as we need to @orta.

**Pull request merged**